### ISSUE TITLE
Spack cannot do anything from clean install

### ISSUE THREAD

**Issue by mathomp4** [LLNL/spack#44434]
### Steps to reproduce

Starting with a clean Ubuntu instance on AWS, I can't seem to get spack to do anything. Now, I thought what I'd done is reproduce #44145 so I did `apt install binutils` since I didn't even have `ld` on my system. But, that doesn't seem to do it. So, to wit:

```console
$ mkdir empty
$ cd empty
$ git clone -c feature.manyFiles=true https://github.com/spack/spack.git
...
$ . spack/share/spack/setup-env.sh
```

After that:
```console
$ spack spec gmake
==> Fetching https://mirror.spack.io/bootstrap/github-actions/v0.5/build_cache/linux-centos7-x86_64-gcc-10.2.1-clingo-bootstrap-spack-zcw5ieomfwwpzpzpabetix2plfqzpvwd.spec.json
==> Fetching https://mirror.spack.io/bootstrap/github-actions/v0.5/build_cache/linux-centos7-x86_64/gcc-10.2.1/clingo-bootstrap-spack/linux-centos7-x86_64-gcc-10.2.1-clingo-bootstrap-spack-zcw5ieomfwwpzpzpabetix2plfqzpvwd.spack
==> Installing "clingo-bootstrap@=spack%gcc@=10.2.1~docs+ipo+optimized+python+static_libstdcpp build_system=cmake build_type=Release generator=make patches=bebb819,ec99431 arch=linux-centos7-x86_64" from a buildcache
/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1081:6-41: info: atom does not occur in any rule head:
  compiler_name(CompilerID,Compiler)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1082:6-43: info: atom does not occur in any rule head:
  compiler_version(CompilerID,Version)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1118:44-67: info: atom does not occur in any rule head:
  compiler_id(CompilerID)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1126:6-45: info: atom does not occur in any rule head:
  compiler_name(CompilerID,CompilerName)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1127:6-51: info: atom does not occur in any rule head:
  compiler_version(CompilerID,CompilerVersion)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1133:6-45: info: atom does not occur in any rule head:
  compiler_name(CompilerID,CompilerName)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1134:6-51: info: atom does not occur in any rule head:
  compiler_version(CompilerID,CompilerVersion)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1173:6-41: info: atom does not occur in any rule head:
  compiler_name(CompilerID,Compiler)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1194:6-41: info: atom does not occur in any rule head:
  compiler_name(CompilerID,Compiler)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1195:6-43: info: atom does not occur in any rule head:
  compiler_version(CompilerID,Version)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1196:6-41: info: atom does not occur in any rule head:
  compiler_os(CompilerID,CompilerOS)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1222:5-40: info: atom does not occur in any rule head:
  compiler_name(CompilerID,Compiler)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1223:5-36: info: atom does not occur in any rule head:
  compiler_version(CompilerID,V)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1224:5-40: info: atom does not occur in any rule head:
  compiler_weight(CompilerID,Weight)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1228:5-40: info: atom does not occur in any rule head:
  compiler_name(CompilerID,Compiler)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1229:5-36: info: atom does not occur in any rule head:
  compiler_version(CompilerID,V)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1230:9-39: info: atom does not occur in any rule head:
  compiler_weight(#X0,#P1)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1254:6-29: info: atom does not occur in any rule head:
  compiler_id(CompilerID)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1289:6-29: info: atom does not occur in any rule head:
  compiler_id(CompilerID)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1290:6-45: info: atom does not occur in any rule head:
  compiler_name(CompilerID,CompilerName)

==> Error: Spack concretizer internal error. Please submit a bug report and include the command, environment if applicable and the following error message.
    gmake is unsatisfiable
```

### Error message

```console
$ spack --debug --stacktrace spec gmake
lib/spack/spack/config.py:1123 ==> [2024-05-29-15:17:53.521471] Reading config from file /home/ubuntu/empty/spack/etc/spack/defaults/config.yaml
lib/spack/spack/cmd/__init__.py:119 ==> [2024-05-29-15:17:53.541978] Imported spec from built-in commands
lib/spack/spack/cmd/__init__.py:119 ==> [2024-05-29-15:17:53.542530] Imported spec from built-in commands
lib/spack/spack/database.py:646 ==> [2024-05-29-15:17:53.543526] DATABASE LOCK TIMEOUT: 60s
lib/spack/spack/store.py:181 ==> [2024-05-29-15:17:53.543652] PACKAGE LOCK TIMEOUT: No timeout
lib/spack/spack/database.py:646 ==> [2024-05-29-15:17:53.544966] DATABASE LOCK TIMEOUT: 60s
lib/spack/spack/store.py:181 ==> [2024-05-29-15:17:53.545102] PACKAGE LOCK TIMEOUT: No timeout
lib/spack/spack/config.py:1123 ==> [2024-05-29-15:17:53.545313] Reading config from file /home/ubuntu/empty/spack/etc/spack/defaults/concretizer.yaml
lib/spack/spack/config.py:1123 ==> [2024-05-29-15:17:53.550069] Reading config from file /home/ubuntu/empty/spack/etc/spack/defaults/packages.yaml
lib/spack/spack/config.py:1123 ==> [2024-05-29-15:17:53.571539] Reading config from file /home/ubuntu/empty/spack/etc/spack/defaults/repos.yaml
lib/spack/spack/config.py:1123 ==> [2024-05-29-15:17:53.654457] Reading config from file /home/ubuntu/.spack/linux/compilers.yaml
lib/spack/spack/database.py:646 ==> [2024-05-29-15:17:53.860488] DATABASE LOCK TIMEOUT: 60s
lib/spack/spack/store.py:181 ==> [2024-05-29-15:17:53.860658] PACKAGE LOCK TIMEOUT: No timeout
lib/spack/spack/config.py:1123 ==> [2024-05-29-15:17:53.861899] Reading config from file /home/ubuntu/empty/spack/etc/spack/defaults/mirrors.yaml
lib/spack/spack/config.py:1123 ==> [2024-05-29-15:17:53.865862] Reading config from file /home/ubuntu/empty/spack/etc/spack/defaults/bootstrap.yaml
lib/spack/spack/config.py:1123 ==> [2024-05-29-15:17:53.871146] Reading config from file /home/ubuntu/empty/spack/etc/spack/defaults/repos.yaml
lib/spack/spack/bootstrap/config.py:125 ==> [2024-05-29-15:17:53.873203] [BOOTSTRAP CONFIG SCOPE] name=_builtin
lib/spack/spack/config.py:1123 ==> [2024-05-29-15:17:53.873601] Reading config from file /home/ubuntu/empty/spack/etc/spack/defaults/bootstrap.yaml
lib/spack/spack/bootstrap/config.py:138 ==> [2024-05-29-15:17:53.878468] [BOOTSTRAP CONFIG SCOPE] name=defaults, path=/home/ubuntu/empty/spack/etc/spack/defaults
lib/spack/spack/bootstrap/config.py:139 ==> [2024-05-29-15:17:53.878621] [BOOTSTRAP CONFIG SCOPE] name=defaults/linux, path=/home/ubuntu/empty/spack/etc/spack/defaults/linux
lib/spack/spack/bootstrap/config.py:138 ==> [2024-05-29-15:17:53.878764] [BOOTSTRAP CONFIG SCOPE] name=bootstrap, path=/home/ubuntu/.spack/bootstrap/config
lib/spack/spack/bootstrap/config.py:139 ==> [2024-05-29-15:17:53.878893] [BOOTSTRAP CONFIG SCOPE] name=bootstrap/linux, path=/home/ubuntu/.spack/bootstrap/config/linux
lib/spack/spack/config.py:1123 ==> [2024-05-29-15:17:53.879714] Reading config from file /home/ubuntu/empty/spack/etc/spack/defaults/config.yaml
lib/spack/spack/config.py:1123 ==> [2024-05-29-15:17:53.892474] Reading config from file /home/ubuntu/.spack/bootstrap/config/linux/config.yaml
lib/spack/spack/database.py:646 ==> [2024-05-29-15:17:53.899895] DATABASE LOCK TIMEOUT: 60s
lib/spack/spack/store.py:181 ==> [2024-05-29-15:17:53.900080] PACKAGE LOCK TIMEOUT: No timeout
lib/spack/spack/config.py:1123 ==> [2024-05-29-15:17:53.900361] Reading config from file /home/ubuntu/empty/spack/etc/spack/defaults/packages.yaml
lib/spack/spack/config.py:1123 ==> [2024-05-29-15:17:53.920878] Reading config from file /home/ubuntu/.spack/bootstrap/config/linux/compilers.yaml
lib/spack/spack/bootstrap/_common.py:228 ==> [2024-05-29-15:17:54.072058] [BOOTSTRAP ROOT SPEC] clingo-bootstrap@spack+python %gcc target=x86_64
lib/spack/spack/bootstrap/core.py:355 ==> [2024-05-29-15:17:54.072201] [BOOTSTRAP MODULE clingo] Try importing from Python
lib/spack/spack/config.py:1123 ==> [2024-05-29-15:17:54.072492] Reading config from file /home/ubuntu/empty/spack/etc/spack/defaults/bootstrap.yaml
lib/spack/spack/bootstrap/_common.py:83 ==> [2024-05-29-15:17:54.124372] [BOOTSTRAP MODULE clingo] The installed spec "clingo-bootstrap@spack+python %gcc target=x86_64 ^python@3.12/zcw5ieomfwwpzpzpabetix2plfqzpvwd" provides the "clingo" Python module
lib/spack/spack/config.py:1123 ==> [2024-05-29-15:17:54.124924] Reading config from file /home/ubuntu/empty/spack/etc/spack/defaults/concretizer.yaml
lib/spack/spack/config.py:1123 ==> [2024-05-29-15:17:54.135804] Reading config from file /home/ubuntu/empty/spack/etc/spack/defaults/packages.yaml
lib/spack/spack/config.py:1123 ==> [2024-05-29-15:17:54.550627] Reading config from file /home/ubuntu/.spack/linux/compilers.yaml
lib/spack/spack/config.py:1123 ==> [2024-05-29-15:17:54.703322] Reading config from file /home/ubuntu/empty/spack/etc/spack/defaults/config.yaml
/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1081:6-41: info: atom does not occur in any rule head:
  compiler_name(CompilerID,Compiler)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1082:6-43: info: atom does not occur in any rule head:
  compiler_version(CompilerID,Version)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1118:44-67: info: atom does not occur in any rule head:
  compiler_id(CompilerID)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1126:6-45: info: atom does not occur in any rule head:
  compiler_name(CompilerID,CompilerName)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1127:6-51: info: atom does not occur in any rule head:
  compiler_version(CompilerID,CompilerVersion)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1133:6-45: info: atom does not occur in any rule head:
  compiler_name(CompilerID,CompilerName)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1134:6-51: info: atom does not occur in any rule head:
  compiler_version(CompilerID,CompilerVersion)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1173:6-41: info: atom does not occur in any rule head:
  compiler_name(CompilerID,Compiler)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1194:6-41: info: atom does not occur in any rule head:
  compiler_name(CompilerID,Compiler)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1195:6-43: info: atom does not occur in any rule head:
  compiler_version(CompilerID,Version)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1196:6-41: info: atom does not occur in any rule head:
  compiler_os(CompilerID,CompilerOS)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1222:5-40: info: atom does not occur in any rule head:
  compiler_name(CompilerID,Compiler)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1223:5-36: info: atom does not occur in any rule head:
  compiler_version(CompilerID,V)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1224:5-40: info: atom does not occur in any rule head:
  compiler_weight(CompilerID,Weight)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1228:5-40: info: atom does not occur in any rule head:
  compiler_name(CompilerID,Compiler)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1229:5-36: info: atom does not occur in any rule head:
  compiler_version(CompilerID,V)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1230:9-39: info: atom does not occur in any rule head:
  compiler_weight(#X0,#P1)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1254:6-29: info: atom does not occur in any rule head:
  compiler_id(CompilerID)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1289:6-29: info: atom does not occur in any rule head:
  compiler_id(CompilerID)

/home/ubuntu/empty/spack/lib/spack/spack/solver/concretize.lp:1290:6-45: info: atom does not occur in any rule head:
  compiler_name(CompilerID,CompilerName)

lib/spack/spack/main.py:1072 ==> [2024-05-29-15:17:56.490393] SolverError: Spack concretizer internal error. Please submit a bug report and include the command, environment if applicable and the following error message.
    gmake is unsatisfiable
lib/spack/spack/error.py:52 ==> [2024-05-29-15:17:56.490633] Error: Spack concretizer internal error. Please submit a bug report and include the command, environment if applicable and the following error message.
    gmake is unsatisfiable
Traceback (most recent call last):
  File "/home/ubuntu/empty/spack/lib/spack/spack/main.py", line 1069, in main
    return _main(argv)
           ^^^^^^^^^^^
  File "/home/ubuntu/empty/spack/lib/spack/spack/main.py", line 1022, in _main
    return finish_parse_and_run(parser, cmd_name, args, env_format_error)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/empty/spack/lib/spack/spack/main.py", line 1052, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/empty/spack/lib/spack/spack/main.py", line 649, in _invoke_command
    return_val = command(parser, args)
                 ^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/empty/spack/lib/spack/spack/cmd/spec.py", line 101, in spec
    concretized_specs = spack.cmd.parse_specs(args.specs, concretize=True)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/empty/spack/lib/spack/spack/cmd/__init__.py", line 177, in parse_specs
    spec.concretize(tests=tests)
  File "/home/ubuntu/empty/spack/lib/spack/spack/spec.py", line 2995, in concretize
    self._new_concretize(tests)
  File "/home/ubuntu/empty/spack/lib/spack/spack/spec.py", line 2968, in _new_concretize
    result = solver.solve([self], tests=tests, allow_deprecated=allow_deprecated)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/empty/spack/lib/spack/spack/solver/asp.py", line 3849, in solve
    result, _, _ = self.driver.solve(
                   ^^^^^^^^^^^^^^^^^^
  File "/home/ubuntu/empty/spack/lib/spack/spack/solver/asp.py", line 929, in solve
    result.raise_if_unsat()
  File "/home/ubuntu/empty/spack/lib/spack/spack/solver/asp.py", line 472, in raise_if_unsat
    raise SolverError(constraints, conflicts=conflicts)
spack.solver.asp.SolverError: Spack concretizer internal error. Please submit a bug report and include the command, environment if applicable and the following error message.
    gmake is unsatisfiable
```

### Information on your system

* **Spack:** 0.23.0.dev0 (b9dfae47227a5e49f430632868c623f5065695bd)
* **Python:** 3.12.3
* **Platform:** linux-ubuntu24.04-icelake
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by mathomp4** [comment]
Note this is a *very* basic install of Ubuntu 24.04. The only package I deliberately installed was `binutils` after seeing the interchange between @haampie and @preney in #44145

**Comment by mathomp4** [comment]
Crap I'm an idiot. I didn't see:

https://spack.readthedocs.io/en/latest/getting_started.html#system-prerequisites

Though, my Ubuntu doesn't have `python3-distutils` I guess because I have Python 3.12?

I'll close with my stupidity!

**Comment by preney** [comment]
The fix done in #44145 worked. (I think I might have missed saying such there? My apologies for that.)

Compilers are crucial to Spack so this leads me to wonder exactly what is Spack doing to detect compilers? Clearly if no compilers can be detected nothing will work and there should be a manual recourse (e.g., in the docs and mentioned by Spack). Any issues with that could have documentation on what persons need to do to get their compilers used?

Another related issue that I run into is when the compiler(s) that Spack detected when I installed it are no longer available (e.g., they've been removed/upgraded to newer versions, etc.). Does Spack have a way to remove compilers no longer present in the system outside of Spack? Currently I manually edit the compilers.yaml file... this really should be an automated feature if explicitly chosen.
